src/message_bus_fuzz.zig

```diff
@@ -0,0 +1,914 @@
+                );
+            },
+            .close => |operation| {
+                io.connections_closing -= 1;
+
+                const local = io.connections.getPtr(operation.fd) orelse {
+                    // close() was called but we didn't ever connect.
+                    // (This happens when we injected an error into connect()).
+                    const result: CloseError!void = {};
+                    completion.callback(completion.context, completion, &result);
+                    return .done;
+                };
+
+                local.closed = true;
+                if (io.prng.chance(io.options.close_error_probability)) {
+                    const result: CloseError!void = io.prng.error_uniform(CloseError);
+                    completion.callback(completion.context, completion, &result);
+                } else {
+                    const result: CloseError!void = {};
+                    completion.callback(completion.context, completion, &result);
+                }
+            },
+            .connect => |operation| {
+                if (io.prng.chance(io.options.connect_error_probability)) {
+                    const result: ConnectError!void = io.prng.error_uniform(ConnectError);
+                    completion.callback(completion.context, completion, &result);
+                    return .done;
+                }
+
+                for (io.servers.values()) |*server| {
+                    if (server.address.eql(operation.address)) {
+                        try server.backlog.append(gpa, completion);
+                        break;
+                    }
+                } else {
+                    // No one listening at that address.
+                    const result: ConnectError!void = error.ConnectionRefused;
+                    completion.callback(completion.context, completion, &result);
+                    return .done;
+                }
+            },
+            .send => |operation| {
+                const sender = io.connections.getPtr(operation.socket).?;
+                assert(!sender.closed);
+                assert(sender.pending_send);
+                sender.pending_send = false;
+
+                if (io.prng.chance(io.options.send_error_probability)) {
+                    const result: SendError!usize = io.prng.error_uniform(SendError);
+                    completion.callback(completion.context, completion, &result);
+
+                    // If there is a pending recv(), we need to make sure that it fails (rather than
+                    // potentially stalling and preventing MessageBus from calling shutdown/close).
+                    sender.closed = true;
+                    return .done;
+                }
+
+                if (sender.shutdown_send) {
+                    const result: SendError!usize = error.BrokenPipe;
+                    completion.callback(completion.context, completion, &result);
+                    return .done;
+                }
```

Comment:
Hm, can we do it probabilistically? I am not entirely sure, but I think if you call `send`, and then `shutdown`, the `send` might still complete after shutdown?